csbpg/db_dumps_test.go

```diff
@@ -0,0 +1,131 @@
+package csbpg
+
+import (
+	"context"
+	"os/exec"
+	"strings"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/onsi/gomega/gexec"
+)
+
+var _ = Describe("for every supported Postgres and IAAS combination", Label("dumps"), func() {
+	When("we use a dump from GCP's postgres 14", func() { testBindingCommonOps("14", "gcp_pg14.sql") })
+	When("we use a dump from GCP's postgres 15", func() { testBindingCommonOps("15", "gcp_pg15.sql") })
+	When("we use a dump from AWS's postgres 14", func() { testBindingCommonOps("14", "aws_pg14.sql") })
+	When("we use a dump from AWS's postgres 15", func() { testBindingCommonOps("15", "aws_pg15.sql") })
+	When("we use a dump from AWS's aurora postgres 14", func() { testBindingCommonOps("14", "aws_aurora_pg14.sql") })
+	When("we use a dump from AWS's aurora postgres 15", func() { testBindingCommonOps("15", "aws_aurora_pg15.sql") })
+})
+
+func testBindingCommonOps(pgVersion, dumpFile string) {
+	var factory connectionFactory
+	var err error
+
+	BeforeEach(func() {
+		factory, err = preparePostgresInstance(pgVersion, dumpFile)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		err = cleanPostgresInstance(pgVersion, dumpFile)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	It("can perform all common operations with bindings", func() {
+		ctx := context.TODO()
+
+		By("creating a new user", func() {
+			diag := sqlUserCreate(ctx, "someuser", "someuser", factory)
+			Expect(diag).To(BeNil())
+		})
+
+		By("creating a table as the new user", func() {
+			db, err := factory.ConnectAsUser("someuser", "someuser")
+			Expect(err).NotTo(HaveOccurred())
+			defer db.Close()
+			_, err = db.Exec("CREATE TABLE PUBLIC.AAA();")
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		By("creating a second user", func() {
+			diag := sqlUserCreate(ctx, "otheruser", "otheruser", factory)
+			Expect(diag).To(BeNil())
+		})
+
+		By("deleting the first user", func() {
+			diag := sqlUserDelete(ctx, "someuser", "someuser", factory)
+			Expect(diag).To(BeNil())
+		})
+
+		By("reading the table created by the now deleted first user", func() {
+			db, err := factory.ConnectAsUser("otheruser", "otheruser")
+			Expect(err).NotTo(HaveOccurred())
+			defer db.Close()
+			_, err = db.Exec("SELECT * FROM PUBLIC.AAA;")
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		By("failing to read a non existing table", func() {
+			db, err := factory.ConnectAsUser("otheruser", "otheruser")
+			Expect(err).NotTo(HaveOccurred())
+			defer db.Close()
+			_, err = db.Exec("SELECT * FROM PUBLIC.NON-EXISTING;")
+			Expect(err).To(HaveOccurred())
+		})
```

Comment:
Can the new user delete/modify tables created by the first user?

Reply:
Good question! What is the expected behaviour?
I haven't seen any such test to use as a reference.
Not even in acceptance_tests:
https://github.com/cloudfoundry/csb-brokerpak-aws/blob/main/acceptance-tests/postgresql_test.go

Reply:
The way to rotate credentials is to unbind/bind again. Expectation is that new user will have access to exactly the same set of resources and with the same privileges. A very common usage of a database from the application is for it to create/modify the database schema on startup. If the application creates a set of tables initially, rotates credentials unbinding/binding, and after some time needs to add a new field to a table to support a new feature, then the new binding user needs to be able to modify tables created by the original user that created it.
Summarising, the new user must be able to modify tables that were created by a previous user.

Reply:
I didn't know that wasn't tested, I would definitely addd such test.

Reply:
I believe the behaviour you are describing is not working at the moment.

I have created a postgres instance using `csb-brokerpak-aws`'s `main` branch and tested the following:
- Created binding1
- Created Table1 logged as binding1
- Created binding2
- Tried doing a SELECT * FROM Table1 logged as binding2 and it succeeded
- Tried adding a new column to Table1 logged as binding2 but it failed
- Tried adding a new column to Table1 logged as binding1 and it succeeded
- Tried deleting Table1 logged as binding2 but it failed

I'm not sure whether this is some kind of regression.
However I haven't found any tests describing such functionality,
which makes me think that maybe that behaviour was never intended or implemented.

Reply:
What DOES WORK is trying to modify or delete tables created by binding1 AFTER it has been properly deleted using csb.

This behaviour can be traced down to the following lines:
https://github.com/cloudfoundry/terraform-provider-csbpg/blob/351d269dde8a45b66baacb5f018266550bc81042/csbpg/resource_binding_user.go#L204-L208

This isn't the only surprising behaviour. The following is also quite interesting:
- Create binding1
- Create Table1 using binding1
- Create binding2
- SELECT FROM Table1 using binding2 succeeds
- Create Table2 using binding1
- SELECT FROM Table2 using binding2 fails
- Create binding3
- SELECT FROM Table2 using binding2 succeeds
- SELECT FROM Table2 using binding3 succeeds

This behaviour can be traced down to the following line:
https://github.com/cloudfoundry/terraform-provider-csbpg/blob/351d269dde8a45b66baacb5f018266550bc81042/csbpg/data_owner_role.go#L37-L39